src/Nethermind/Nethermind.Synchronization/Trie/NodeDataRecovery.cs

```diff
@@ -0,0 +1,192 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Blockchain.Synchronization;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Tasks;
+using Nethermind.Core.Utils;
+using Nethermind.Logging;
+using Nethermind.Network.Contract.P2P;
+using Nethermind.State.Healing;
+using Nethermind.State.Snap;
+using Nethermind.Stats;
+using Nethermind.Synchronization.Peers;
+using Nethermind.Synchronization.Peers.AllocationStrategies;
+using Nethermind.Trie;
+using Nethermind.Trie.Pruning;
+
+namespace Nethermind.Synchronization.Trie;
+
+public class NodeDataRecovery(ISyncPeerPool peerPool, INodeStorage nodeStorage, ILogManager logManager) : IPathRecovery
+{
+    private static readonly IPeerAllocationStrategy NodePeerStrategy =
+        new CanServeByHashPeerAllocationStrategy(
+            new BySpeedStrategy(TransferSpeedType.Latency, false));
+
+    private const int ConcurrentAttempt = 3;
+    private readonly ILogger _logger = logManager.GetClassLogger<NodeDataRecovery>();
+
+    public async Task<IDictionary<TreePath, byte[]>?> Recover(Hash256 rootHash, Hash256? address, TreePath startingPath, Hash256 startingNodeHash, Hash256 fullPath, CancellationToken cancellationToken)
+    {
+        using AutoCancelTokenSource cts = cancellationToken.CreateChildTokenSource();
+
+        Hash256 currentHash = startingNodeHash;
+        TreePath currentPath = startingPath;
+        TreePath queryPath = new TreePath(fullPath, 64);
+
+        // Sometimes the start path for the missing node and the actual full path that the trie is working on is not the same.
+        // So we change the query to match the missing node path.
+        if (queryPath.Truncate(startingPath.Length) != startingPath)
+        {
+            int remainingLength = 64 - startingPath.Length;
+            TreePath newQueryPath = startingPath;
+            for (int i = 0; i < remainingLength; i++)
+            {
+                newQueryPath.Append(0);
+            }
+            queryPath = newQueryPath;
+        }
+
+        Dictionary<TreePath, byte[]> recoveredNodes = new();
```

Comment:
Does it make sense to be `Dictionary`? Wouldn't list suffice? (could use pooledList)